Battery_Mon: UAVCAN based MPPT

magicrub opened the pull request:
This adds a couple things:
- new driver for PacketDigital UAVCAN MPPT
- new DroneCAN msgs (needs PR https://github.com/dronecan/DSDL/pull/3)
- option bits for turning power on/off at boot/arm events For MPPT only

rmackay9 commented:
I have tested this on real hardware an confirmed that it works.
![image](https://user-images.githubusercontent.com/1498098/144157366-abb4d6bb-dade-41ae-95b1-d8a9fe074bac.png)

Also here is an early wiki page for the Packet Digital MPPT board. https://ardupilot.org/rover/docs/common-packetdigital-mppt.html

magicrub commented:
We'll dang, looks like I left a couple debug things in there along with some commented out debug code

tridge commented:
@magicrub I added a comment about the enable/disable bool setup on the DSDL

magicrub commented:
The DroneCAN PR has been merged. The module commit has been updated in this PR.

rmackay9 commented:
I've retested this on a Rover and confirmed that it appears to be working.  I can see the battery voltage being reported and the enable/disable of the MPPT output as the vehicle arms/disarms seems to be working as well.
![image](https://user-images.githubusercontent.com/1498098/144344185-3a1360be-fad1-45ba-8484-842a9bfb20c8.png)

magicrub commented on libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp:
```diff
@@ -253,3 +253,28 @@ bool AP_BattMonitor_Backend::reset_remaining(float percentage)
+void AP_BattMonitor_Backend::set_bootup_powerstate()
+{
+    const uint32_t options = uint32_t(_params._options.get());
+    const bool on_at_boot = uint32_t(AP_BattMonitor_Params::Options::Power_ON_at_Boot) != 0;
```
doh! This should be
```suggestion
    const bool on_at_boot = (options & uint32_t(AP_BattMonitor_Params::Options::Power_ON_at_Boot)) != 0;
```

magicrub replied:
fixed

rmackay9 commented on libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:
```diff
@@ -192,16 +267,67 @@ void AP_BattMonitor_UAVCAN::read()
+void AP_BattMonitor_UAVCAN::handle_mppt_enable_output_response(const uavcan::ServiceCallResult<mppt::OutputEnable>& response)
+{
+    // TODO
+    //const mppt::OutputEnable::Response r = response.getResponse();
```
normally we try not to leave commented out code in there..

magicrub replied:
How about an # if 0?

rmackay9 replied:
Are we really ever going to support it?  If the commented out code is going to sit there forever let's just remove it and add an "unsupported" comment or something similar.

magicrub replied:
replaced with "Not Supported"

rmackay9 commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.h:
```diff
@@ -18,7 +18,12 @@ class AP_BattMonitor_Params {
     enum class Options : uint8_t {
-        Ignore_UAVCAN_SoC = (1U<<0),
+        Ignore_UAVCAN_SoC                   = (1U<<0),
+        UAVCAN_MPPT_Use_Input_Value         = (1U<<1),
```
We should add all these bits to the _OPTIONS parameter description

magicrub replied:
Oh, whoops! That must have fallen off at some point. I'll add it back in.

magicrub replied:
thanks for including them in your commit

rmackay9 commented on libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:
```diff
@@ -55,6 +59,14 @@ void AP_BattMonitor_UAVCAN::subscribe_msgs(AP_UAVCAN* ap_uavcan)
+    mppt_stream_listener = new uavcan::Subscriber<mppt::Stream, MpptStreamCb>(*node);
+    const int mppt_stream_listener_res = mppt_stream_listener->start(MpptStreamCb(ap_uavcan, &handle_mppt_stream_trampoline));
+    if (mppt_stream_listener_res < 0) {
+        AP_HAL::panic("UAVCAN Mppt::Stream subscriber start problem\n\r");
```
Surely we don't want a panic here do we?

magicrub replied:
Good catch! That's as copy paste from an older branch where we were doing it that way here. This new way looks much better.

magicrub replied:
I fixed this and added it to your commit since you're the one who caught it

rmackay9 reviewed (changes requested):
I've put some suggested renaming and comment commits here https://github.com/rmackay9/rmackay9-ardupilot/tree/magicrub-pr-mppt.  Some of these are a bit nitpicky so I don't mind if you don't include them all.

I find the way set_hardware_to_powered_state() is called confusing.  It seems to me that we are mixing together two different features.  One is the existing feature to monitor for the battery reporting that it is powering off.  The other is the new feature where the battery turns on/off when the vehicle is armed/disarmed.  It seems to me that it would be better to move the logic to turn on/off the MPPT down into the UAVCAN library and have it just monitor the vehicle's arming state by itself.  Then if it gets turned off, the existing code will probably notice the next time it receives a status message from the battery..

I think this would reduce the complexity of the AP_BatteryMonitor class and instead keep it down in the UAVCAN library where it is out of view for most users.

magicrub commented:
> I've put some suggested renaming and comment commits here https://github.com/rmackay9/rmackay9-ardupilot/tree/magicrub-pr-mppt. Some of these are a bit nitpicky so I don't mind if you don't include them all.


I actually liked all of them except the last two which I fixed up. I squashed them all into a new commit with you as the author. The options description had two bits swapped and I didn't like that you made the options say MPPT specific. This feature was added generically on purpose. Only MPPT supports it now but I think the hooks in place gives us nice flexibility on other devices we support in the future. @tridge also liked the idea because now scripting can take advantage of those hooks too.


> I find the way set_hardware_to_powered_state() is called confusing. It seems to me that we are mixing together two different features. One is the existing feature to monitor for the battery reporting that it is powering off. The other is the new feature where the battery turns on/off when the vehicle is armed/disarmed. 

I agree, I'm kind of converting the Battery_Monitor library to a Battery_Manager library. I think that's OK though.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:
```diff
@@ -153,6 +143,82 @@ void AP_BattMonitor_UAVCAN::handle_battery_info_aux(const BattInfoAuxCb &cb)
+
+        // update total current drawn since startup
+        if (_interim_state.last_time_micros != 0 && dt < 2000000) {
+            // .0002778 is 1/3600 (conversion to hours)
+            float mah = (float) ((double) _interim_state.current_amps * (double) dt * (double) 0.0000002778f);
```
Just put (1.0 / 36000.0) instead of the magic constant. Or break it out into a const. I realize this is a copy paste, but if we are doing it, then let's try and improve it.

I'm also not sold on the use of doubles before casting back down here, it will be much slower on the F4 boards, and I'm not sure the accuracy wins here either, and if it's really that great we should be doing it for the analog monitor.

magicrub replied:
No objection here. It's a copy/paste from here.

magicrub replied:
this has totally been removed.

rmackay9 replied:
We've left this as-is because it is existing code but I've raised an issue here to resolve it in the future https://github.com/ArduPilot/ardupilot/issues/19452

WickedShell reviewed (changes requested):
I haven't actually made it very far, just a passing comment about the fault reporting really stood out. I'm very much in favor of better fault reporting, but we really should do better then send text as our bailing out to report it. It really doesn't help the rest of the system intelligently handle the situation.

magicrub commented on libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:
```diff
@@ -153,6 +143,82 @@ void AP_BattMonitor_UAVCAN::handle_battery_info_aux(const BattInfoAuxCb &cb)
+    for (uint8_t fault_bit=0x01; fault_bit <= 0x08; fault_bit <<= 1) {
+        // this loop is to generate multiple messages if there are multiple concurrent faults, but also run once if there are no faults
+        if ((fault_bit & flags) != 0 || flags == 0) {
+            const MPPT_FaultFlags err = (flags == 0) ? MPPT_FaultFlags::NONE : (MPPT_FaultFlags)fault_bit;
```
No bits will be set so the if would always be false. Can you suggest a better solution?

magicrub commented:
This has been scaled back to only do the arm/disarm on/off stuff for MPPTs only. If we choose to bring that to all batteries, we can do that in another PR later. For this one, lets focus on getting the new MPPT driver in

magicrub commented on libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:
```diff
@@ -153,6 +143,82 @@ void AP_BattMonitor_UAVCAN::handle_battery_info_aux(const BattInfoAuxCb &cb)
+        // this loop is to generate multiple messages if there are multiple concurrent faults, but also run once if there are no faults
+        if ((fault_bit & flags) != 0 || flags == 0) {
+            const MPPT_FaultFlags err = (flags == 0) ? MPPT_FaultFlags::NONE : (MPPT_FaultFlags)fault_bit;
+            GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Battery %u: Fault: %s", (unsigned)_instance+1, mppt_fault_string(err));
```
added the MAVLink flags

rmackay9 replied:
I worked with TomP on this and we've added reporting of MPPT faults to the mavlink BATTERY_STATUS message.  Thanks for the suggestion.

magicrub commented:
big thanks to @rmackay9 for help testing on this!

rmackay9 reviewed (approved):
I very much like the recent changes to reduce the scope of the changes so they (mostly) only affect the AP_BatteryMonitor_UAVCAN library.

I've retested this and it still works.  Thanks @magicrub for this!

rmackay9 commented:
@WickedShell you're OK with this PR now?  https://github.com/ArduPilot/ardupilot/pull/19404

hendjoshsr71 commented on libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:
```diff
@@ -153,6 +176,30 @@ void AP_BattMonitor_UAVCAN::handle_battery_info_aux(const BattInfoAuxCb &cb)
+    const float current = use_input_value ? cb.msg->input_current : cb.msg->output_current;
+
+    // use an invalid soc so we use the library calculated one
+    const uint8_t soc = 101;
```
Another not for this PR...
Using `soc = 101`  for an invalid soc is pre-existing, but I'm not a fan of it. Technically, of course the battery can never get above 100 soc by definition both in current UAVCAN standard and of course reality. But if the  system is misconfigured (wrong voltages, capacity, etc.) it could end up above 101 SOC.

I just looked some code for a  UAVCAN battery monitor  and it just sends out 100 if it computes SOC above 100. Instead IMO a device should send a "hey this is isn't right state using  127". Just thought as we look to future DroneCan stuff.

rmackay9 replied:
I'm also not a fan of setting SOC=101 as the way to fallback to the calculated SOC.  At least there is a comment to clarify why it is done.

magicrub replied:
It's unused anyway. Where we fetch it it gets ignored if we're an MPPT. I could set to 0? 127?

magicrub replied:
Consider it for a future improvement and just merge now?

hendjoshsr71 replied:
For future.... certainly. Might as well keep 101 since then you have to change it other places too...

magicrub replied:
Ok.. merging!

rmackay9 commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -689,6 +689,9 @@ MAV_BATTERY_CHARGE_STATE AP_BattMonitor::get_mavlink_charge_state(const uint8_t
     switch (state[instance].failsafe) {
 
     case Failsafe::None:
+        if (get_mavlink_fault_bitmask(instance) != 0 || !healthy()) {
```
@hendjoshsr71, I wonder if #18508 is the right PR..?

hendjoshsr71 replied:
@rmackay9, opps i mixed up two numbers here you go... https://github.com/ArduPilot/ardupilot/pull/18058

magicrub commented:
You know.. since the scope got pulled way back, maybe it's worth backporting to plane v4.1.5?
